cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -1479,39 +1532,69 @@ void get_stack_context(device_span<SymbolT const> json_in,
-  // StructBegin, StructEnd. Also, all LineEnd are removed as well, as these are not relevant after
-  // this stage anymore
-  filter_fst.Transduce(
-    thrust::make_reverse_iterator(thrust::make_zip_iterator(tokens.data(), token_indices.data()) +
-                                  tokens.size()),
-    static_cast<SymbolOffsetT>(tokens.size()),
-    thrust::make_reverse_iterator(
-      thrust::make_zip_iterator(filtered_tokens_out.data(), filtered_token_indices_out.data()) +
-      tokens.size()),
-    thrust::make_discard_iterator(),
-    d_num_selected_tokens.data(),
-    token_filter::start_state,
-    stream);
+  if (remove_line_end_token) {
+    // The FST is run on the reverse token stream, discarding all tokens between ErrorBegin and the
+    // next LineEnd (LineEnd, inv_token_0, inv_token_1, ..., inv_token_n, ErrorBegin, LineEnd, ...),
+    // emitting a [StructBegin, StructEnd] pair on the end of such an invalid line. In that example,
+    // inv_token_i for i in [0, n] together with the ErrorBegin are removed and replaced with
+    // StructBegin, StructEnd. Also, all LineEnd are removed as well, as these are not relevant
+    // after this stage anymore
+
+    auto filter_fst = fst::detail::make_fst(
+      fst::detail::make_symbol_group_lut(token_filter::symbol_groups, sgid_op),
+      fst::detail::make_transition_table(token_filter::transition_table),
+      fst::detail::make_translation_functor(token_filter::TransduceToken{}),
+      stream);
+
+    filter_fst.Transduce(
+      thrust::make_reverse_iterator(thrust::make_zip_iterator(tokens.data(), token_indices.data()) +
+                                    tokens.size()),
+      static_cast<SymbolOffsetT>(tokens.size()),
+      thrust::make_reverse_iterator(
+        thrust::make_zip_iterator(filtered_tokens_out.data(), filtered_token_indices_out.data()) +
+        tokens.size()),
+      thrust::make_discard_iterator(),
+      d_num_selected_tokens.data(),
+      token_filter::start_state,
+      stream);
+  } else {
+    // The FST is run on the reverse token stream, discarding all tokens between ErrorBegin and the
+    // next LineEnd (LineEnd, inv_token_0, inv_token_1, ..., inv_token_n, ErrorBegin, LineEnd, ...),
+    // emitting a [LineEnd, StructBegin, StructEnd] on the end of such an invalid line. In that
+    // example, inv_token_i for i in [0, n] together with the ErrorBegin are removed and replaced
+    // with LineEnd, StructBegin, StructEnd. Unlike the previous case, LineEnd tokens are retained
+    // however, the corresponding token index is written as 0.
+
+    auto filter_fst = fst::detail::make_fst(
+      fst::detail::make_symbol_group_lut(token_filter::symbol_groups, sgid_op),
+      fst::detail::make_transition_table(token_filter::transition_table),
+      fst::detail::make_translation_functor(token_filter::TransduceTokenKeepLineEnd{}),
+      stream);
+
+    filter_fst.Transduce(
```

Comment:
This adds 1 more kernel to compile. It will increase both binary size and compile time. Please check compile time of this file in `cpp/build/latest/.ninja_log`.
Ideally if avoidable, prefer adding the boolean as functor member, rather than new functor.

Reply:
> prefer adding the boolean as functor member, rather than new functor.

This is also a good idea.

We want to have the `LineEnd` token so we can easily identify the individual rows. Yes, currently we are doing so using `StructBegin`/`StructEnd` but that requires several more kernel calls (prefix sum then zero check). If we can defer filtering out the `LineEnd` token then it is better.

Reply:
Done!